server/src/main/java/org/elasticsearch/search/retriever/KnnRetrieverBuilder.java

```diff
@@ -0,0 +1,122 @@
+ */
+public final class KnnRetrieverBuilder extends RetrieverBuilder {
+
+    public static final String NAME = "knn";
+    public static final NodeFeature NODE_FEATURE = new NodeFeature(NAME + "_retriever");
+
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField K_FIELD = new ParseField("k");
+    public static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField QUERY_VECTOR_BUILDER_FIELD = new ParseField("query_vector_builder");
+    public static final ParseField VECTOR_SIMILARITY = new ParseField("similarity");
+
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<KnnRetrieverBuilder, RetrieverParserContext> PARSER = new ConstructingObjectParser<>(
+        "knn",
+        args -> {
+            List<Float> vector = (List<Float>) args[1];
+            final float[] vectorArray;
+            if (vector != null) {
+                vectorArray = new float[vector.size()];
+                for (int i = 0; i < vector.size(); i++) {
+                    vectorArray[i] = vector.get(i);
+                }
+            } else {
+                vectorArray = null;
+            }
+            return new KnnRetrieverBuilder(
+                (String) args[0],
+                vectorArray,
+                (QueryVectorBuilder) args[2],
+                (int) args[3],
+                (int) args[4],
+                (Float) args[5]
+            );
+        }
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), FIELD_FIELD);
+        PARSER.declareFloatArray(optionalConstructorArg(), QUERY_VECTOR_FIELD);
+        PARSER.declareNamedObject(
+            optionalConstructorArg(),
+            (p, c, n) -> p.namedObject(QueryVectorBuilder.class, n, c),
+            QUERY_VECTOR_BUILDER_FIELD
+        );
+        PARSER.declareInt(constructorArg(), K_FIELD);
+        PARSER.declareInt(constructorArg(), NUM_CANDS_FIELD);
+        PARSER.declareFloat(optionalConstructorArg(), VECTOR_SIMILARITY);
+        RetrieverBuilder.declareBaseParserFields(NAME, PARSER);
+    }
+
+    public static KnnRetrieverBuilder fromXContent(XContentParser parser, RetrieverParserContext context) throws IOException {
+        if (context.clusterSupportsFeature(NODE_FEATURE) == false) {
+            throw new ParsingException(parser.getTokenLocation(), "unknown retriever [" + NAME + "]");
+        }
+        return PARSER.apply(parser, context);
+    }
+
+    private final String field;
+    private final float[] queryVector;
+    private final QueryVectorBuilder queryVectorBuilder;
+    private final int k;
+    private final int numCands;
+    private final Float similarity;
+
+    public KnnRetrieverBuilder(
+        String field,
+        float[] queryVector,
+        QueryVectorBuilder queryVectorBuilder,
+        int k,
```

Comment:
`k` and `numCand` became [optional in Knn search](https://github.com/elastic/elasticsearch/pull/101209), I wonder if should make it optional too for retriever?

Reply:
I would strongly prefer to do this as a follow up.

Reply:
A bit late here, but in addition to this we're also changing `queryVector` parsing to a generate a `VectorData` object instead of directly parsing to `float[]` (https://github.com/elastic/elasticsearch/pull/105393). This should probably be  part of the follow up  work as well.